### PR TITLE
Position tablet relative to hand

### PR DESCRIPTION
**Pull request by ctrlaltdavid**

### CODE REVIEW

**Comment by ctrlaltdavid** [comment]
Testing

Test the orientation that the tablet is rezzed in HMD mode using the controllers, for various hand positions, both left and right hands.
Test subsequently using the tablet with stylus and with finger.

Test the orientation that the tablet is rezzed in HMD mode using a mouse.

**Comment by ctrlaltdavid** [comment]
For Piper to QA.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/140/HighFidelity-Beta-PR10538-39654779fbfa9ca1015b65c3595623877f7921b4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/140/HighFidelity-Beta-PR10538-39654779fbfa9ca1015b65c3595623877f7921b4.dmg

**Comment by PhilipRosedale** [comment]
@ctrlaltdavid @Nex-Pro I like this a lot in terms of the leanback and being right at and in front of my hand.  But can we make it always vertical?  Coming up with the icons tilted in the horizontal plane of my vision doesn't feel right.

**Comment by Nex-Pro** [comment]
Testing this.

**Comment by Nex-Pro** [comment]
@PhilipRosedale You gotta stop treating this like a punch. Rolling your wrist thumb down as you put your arm forward isn't helping.  ;)

**Comment by Nex-Pro** [comment]
For me this is a Pass
@PhilipRosedale Are thinking of having the tablet roll match the view frustrum roll ?

**Comment by ctrlaltdavid** [comment]
@PhilipRosedale , @Nex-Pro I've made it so that that tablet is always horizontal w.r.t. the world.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/154/HighFidelity-Beta-PR10538-925cca97db51bbfabd30160d15259c0540208591.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/154/HighFidelity-Beta-PR10538-925cca97db51bbfabd30160d15259c0540208591.dmg

**Comment by Nex-Pro** [comment]
re-test

**Comment by Nex-Pro** [comment]
### FAIL

If the hand is tipped forward slightly to get the tablet leaning back it can suddenly appear to be upside down  there's a point where a change of 1 degree of the hand introduces a change in tilt of the tablet of more than 100 degrees. Also the tablet sometimes appears tilted towards the user in an unpredictable manner.

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/157/HighFidelity-Beta-PR10538-f1dca5019a6f79209b167102cd61cdfc12312855.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/157/HighFidelity-Beta-PR10538-f1dca5019a6f79209b167102cd61cdfc12312855.dmg

**Comment by Nex-Pro** [comment]
retesting this

**Comment by Nex-Pro** [comment]
Noticed is that if the hand is tilted forward sufficiently to tilt the top of the tablet down and away from the avatar a rotation in the world Y axis is introduced (clockwise looking down) This rotation increases with, and seems to be proportional to, the hand controller tilt. However there isn't really a use for tilting the tablet that far ( as far as I can see ).

@PhilipRosedale  I thought I wasn't going to like forcing the tablet horizontal to the ground but this works pretty good apart from... 

If the user's down direction ( eg. walking on the ceiling, in a plane that's turning )  doesn't correspond with that of the local situation in the domain the tablet still appears oriented with regard to the domain.  IMO this needs to work relative to the avatar's down direction.  I don't think this is difficult to implement. Maybe grabbing the avatar's rotation using the quat gleaned from MyAvatar.orientation  rather than world horizontal would do the trick.  

Not due to this PR:
Testing this with my avatar slowly rotating, via a script I wrote, the tablet  rotation doesn't follow the user. This means that if a user is in a vehicle that is turning/banking and opens the tablet it is simply going to be left behind.

**Comment by Nex-Pro** [comment]
@ctrlaltdavid ^

**Comment by ctrlaltdavid** [comment]
I'll have a look at handling users with arbitrary verticals.

**Comment by ctrlaltdavid** [comment]
@Piper notes:
- "if the hand is tilted forward sufficiently" = if the hand is pointing straight down, the "forward" direction in your avatar's local horizontal plane will  change significantly with minor changes to the "down" orientation, hence the tablet rez direction appearing to rotate. Same happens if you point straight up. Nothing to be done about this.
- Avatar orientation is now handled.
- Tablet getting left behind happens in master, too. Though it catches up if you teleport.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/227/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by ctrlaltdavid** [comment]
Build this please

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/228/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/230/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by ctrlaltdavid** [comment]
Test this please

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/237/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by Nex-Pro** [comment]
Testing this

**Comment by ctrlaltdavid** [comment]
@Nex-Pro Windows builds don't seem to be very reliable at present; may not get a build until tomorrow

**Comment by marko8904** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/240/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/240/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by PhilipRosedale** [comment]
Looks good to me, although not exactly what I was thinking.  @Nex-Pro what do you think?

**Comment by themelissabrown** [comment]
@Nex-Pro can you re-test and let me know the results?

**Comment by PhilipRosedale** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/531/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/531/HighFidelity-Beta-PR10538-367b3ec603c889021c23601db78e0fe11bfd4047.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Testing

Test the orientation that the tablet is rezzed in HMD mode using the controllers, for various hand positions, both left and right hands. **Pass** all seems normal. The only note I would make is if I tilt the controller towards me so the vives circular end is facing my head in real life, it will bring up the tablet backwards. 
Test subsequently using the tablet with stylus and with finger. **Pass** works as it would normally.

Test the orientation that the tablet is rezzed in HMD mode using a mouse. **Pass**

Test with non-vertical avatar orientation in HMD:
In desktop mode, in the console issue the command, MyAvatar.orientation = Quat.fromPitchYawRollDegrees(0, 0, -90);
Switch to HMD mode. The world will be angled at 90 degrees.
The tablet should rez with horizontal per your avatar's "horizontal".
When rotated 90 degrees:
image
**Pass?** I don't seem to notice anything specifically wrong with this.

Please let me know if there is anything more I can do/provide to assist. :)

File: scripts/system/libraries/WebTablet.js [hunk 1]
```diff
@@ -51,30 +51,33 @@ function calcSpawnInfo(hand, height) {
...
+        var rotation = handController.rotation;
+
+        if (hand === Controller.Standard.LeftHand) {
+            rotation = Quat.multiply(rotation, Quat.fromPitchYawRollDegrees(0, 90, 0));
+        } else {
+            rotation = Quat.multiply(rotation, Quat.fromPitchYawRollDegrees(0, -90, 0));
+        }
+        var normal = Vec3.multiplyQbyV(rotation, Vec3.UNIT_NEG_Y);
+        var lookAt =  Quat.lookAt({x: 0, y: 0, z: 0}, normal, Vec3.multiplyQbyV(MyAvatar.orientation, Vec3.UNIT_Y));
+        rotation = Quat.multiply(Quat.angleAxis(30, Vec3.multiplyQbyV(lookAt, Vec3.UNIT_X)), lookAt);
```
Comment (Atlante45) [inline scripts/system/libraries/WebTablet.js]: Magic number

File: scripts/system/libraries/WebTablet.js [hunk 2]
```diff
@@ -51,30 +51,33 @@ function calcSpawnInfo(hand, height) {
...
 
-        var position = Vec3.sum(handController.position, {x: 0, y: TABLET_UP_OFFSET, z: 0});
-        var rotation = Quat.lookAt({x: 0, y: 0, z: 0}, normal, Y_AXIS);
-        var offset = Vec3.multiplyQbyV(rotation, {x: 0, y: height / 2, z: TABLET_FORWARD_OFFSET});
+        position = Vec3.sum(position, Vec3.multiplyQbyV(rotation, { x: 0, y: tabletHeight * 0.4, z: tabletHeight * 0.05 }));
```
Comment (Atlante45) [inline scripts/system/libraries/WebTablet.js]: Magic numbers

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/564/HighFidelity-Beta-PR10538-fff8876b21154e7c260d4929e017099a88bec45c.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/564/HighFidelity-Beta-PR10538-fff8876b21154e7c260d4929e017099a88bec45c.dmg

**Comment by ctrlaltdavid** [comment]
Quick retest needed before merge in case code changes introduced bug. (I don't have HMD to test with today.)

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete: 

All acted the same as before in previous build of this PR, **Passes** all around :D

**Pull request merged**